src/StrokeEngine.cpp

```diff
@@ -1,551 +1,177 @@
-        }
-    }
+  // Initialize with default values
+  motionBounds bounds = motor->getBounds();
+
+  this->maxDepth = abs(bounds.end - bounds.start);
+  this->depth = this->maxDepth; 
+  this->stroke = this->maxDepth / 3;
+  this->strokeRate = 1.0;
+  this->sensation = 0.0;
+
+  ESP_LOGD("StrokeEngine", "Stroke Parameter Max Depth = %f", this->maxDepth);
+  ESP_LOGD("StrokeEngine", "Stroke Parameter Depth = %f", this->depth);
+  ESP_LOGD("StrokeEngine", "Stroke Parameter Stroke = %f", this->stroke);
+  ESP_LOGD("StrokeEngine", "Stroke Parameter Stroke Rate = %f", this->strokeRate);
+  ESP_LOGD("StrokeEngine", "Stroke Parameter Sensation = %f", this->sensation);
+
+  ESP_LOGI("StrokeEngine", "Attached Motor succesfully to Stroke Engine!");
+}
+
+void StrokeEngine::setParameter(StrokeParameter parameter, float value, bool applyNow) {
+  if (xSemaphoreTake(_patternMutex, portMAX_DELAY) != pdTRUE) {
+    return;
+  }
+
+  String name;
+  float debugValue;
+
+  switch (parameter) {
+    // TODO - Add SPEED
+    // TODO - When rate is set to 1 it bugs out and stops all motion
+    case StrokeParameter::RATE:
+      name = "Stroke Rate";
+      debugValue = this->strokeRate = constrain(value, 1, 60 * 6);
+      break;
     
-#ifdef DEBUG_TALKATIVE
-    Serial.println("Stroke Engine State: " + verboseState[_state]);
-#endif
-}
-
-void StrokeEngine::enableAndHome(endstopProperties *endstop, void(*callBackHoming)(bool), float speed) {
-    // Store callback
-    _callBackHomeing = callBackHoming;
-
-    // enable and home
-    enableAndHome(endstop, speed);
+    case StrokeParameter::DEPTH:
+      name = "Depth";
+      debugValue = this->depth = constrain(int(value), 0, this->maxDepth); 
+      break;
+
+    case StrokeParameter::STROKE:
+      name = "Stroke";
+      debugValue = this->stroke = constrain(int(value), 0, this->maxDepth); 
+      break;
+
+    case StrokeParameter::SENSATION:
+      name = "Sensation";
+      debugValue = this->sensation = constrain(sensation, -100, 100); 
+      break;
+
+    case StrokeParameter::PATTERN:
+      name = "Pattern";
+      debugValue = this->_patternIndex = value;
+      this->_index = 0;
+      break;
+  }
+  
+  this->sendParameters(this->_patternIndex);
+  
+  ESP_LOGD("StrokeEngine", "Stroke Parameter %s - %f", name, debugValue);
+  
+  // When running a pattern and immediate update requested: 
+  if (applyNow == true) {
+    this->applyUpdate = true;
+    ESP_LOGD("StrokeEngine", "Setting Apply Update Flag!");
+  }
+
+  xSemaphoreGive(_patternMutex);
+}
+
+// TODO - Better naming?
+void StrokeEngine::sendParameters(int patternIndex) {
+  // Stroke is dynamically constrainted based on depth.
+  // This allows depth to change, and allow stroke to fill the available space, 
+  // rather than having to set the parameter again
+
+  patternTable[patternIndex]->setSpeedLimit(
+    this->motor->getMaxSpeed(), 
+    this->motor->getMaxAcceleration(),
+    1
+  );
+

```

Comment:
This must be safeguarded by the mutex. Already got a race condition at this point.

Reply:
This is a protected method, and all calls internally wrap a mutex around it. This method will never be used externally, and should be safe. I'll add a comment to it that it does require using a mutex though